cpp/src/arrow/type.h

```diff
@@ -1398,6 +1401,167 @@ class ARROW_EXPORT DictionaryUnifier {
                            std::shared_ptr<Array>* out_dict) = 0;
 };
 
+// ----------------------------------------------------------------------
+// FieldRef
+
+/// \class FieldRef
+/// \brief Descriptor of a (potentially nested) field within a schema.
+class ARROW_EXPORT FieldRef {
+ public:
+  /// Represents a path to a nested field using indices of child fields.
+  /// For example, given indices {5, 9, 3} the field would be retrieved with
+  /// schema->field(5)->type()->child(9)->type()->child(3)
+  ///
+  /// std::basic_string is used to take advantage
+  /// of the short string approximation: up to four indices can be stored without memory
+  /// allocation.
+  using Indices = std::basic_string<int>;
+
+  /// Construct a FieldRef using a string of indices. The reference will be retrieved as:
+  /// schema.fields[self.indices[0]].type.fields[self.indices[1]] ...
+  ///
+  /// Empty indices are not valid.
+  FieldRef(Indices indices);  // NOLINT runtime/explicit
+
+  /// Construct a by-name FieldRef. Multiple fields may match a by-name FieldRef:
+  /// [f for f in schema.fields where f.name == self.name]
+  FieldRef(std::string name) : impl_(std::move(name)) {}  // NOLINT runtime/explicit
+
+  /// Equivalent to a single index string of indices.
+  FieldRef(int index) : impl_(Indices({index})) {}  // NOLINT runtime/explicit
+
+  /// Construct a nested FieldRef.
+  explicit FieldRef(std::vector<FieldRef> children);
+
+  /// Convenience constructor for nested FieldRefs: each argument will be used to
+  /// construct a FieldRef
+  template <typename A0, typename A1, typename... A>
+  FieldRef(A0&& a0, A1&& a1, A&&... a)
+      : FieldRef(std::vector<FieldRef>{
+            // cpplint things the following are constructor decls
+            FieldRef(std::forward<A0>(a0)),       // NOLINT runtime/explicit
+            FieldRef(std::forward<A1>(a1)),       // NOLINT runtime/explicit
+            FieldRef(std::forward<A>(a))...}) {}  // NOLINT runtime/explicit
+
+  /// Parse a dot path into a FieldRef.
+  ///
+  /// dot_path = '.' name
+  ///          | '[' digit+ ']'
+  ///          | dot_path+
+  ///
+  /// Examples:
+  ///   ".alpha" => FieldRef("alpha")
+  ///   "[2]" => FieldRef(2)
+  ///   ".beta[3]" => FieldRef("beta", 3)
+  ///   "[5].gamma.delta[7]" => FieldRef(5, "gamma", "delta", 7)
+  ///   ".hello world" => FieldRef("hello world")
+  ///   R"(.\[y\]\\tho\.\)" => FieldRef(R"([y]\tho.\)")
```

Comment:
Does this example imply `FieldRef` supports regular expressions?

Reply:
No, just that `\` will cause whatever character follows it to be interpreted literally in case your field name must include `\`, '[', ']', or '.'

I'll update the doccomment